compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,320 @@
+        if borrowed_locals.contains(discr.local) {
+            debug!("The aggregate: {discr:?} was borrowed before its discriminant was read");
+            continue;
+        }
+
+        let Location { block, statement_index } = location;
+        let mut finder = MutatedLocalFinder { local: discr.local, mutated: false };
+        for (statement_index, statement) in body.basic_blocks[block]
+            .statements
+            .iter()
+            .enumerate()
+            .take_while(|&(index, _)| index != statement_index)
+        {
+            finder.visit_statement(statement, Location { block, statement_index });
+        }
+
+        if finder.mutated {
+            debug!("The aggregate: {discr:?} was mutated before its discriminant was read");
+            continue;
+        }
+
+        // If `switched` is borrowed by the time we actually switch on it, we also cannot optimize.
+        borrowed_locals.seek_to_block_end(block);
+        if borrowed_locals.contains(switched) {
+            debug!("The local: {switched:?} was borrowed before being switched on");
+            continue;
+        }
+
+        discriminants.insert(
+            switched,
+            Discriminant {
+                block,
+                discr: discr.local,
+                exclude: if ssa_locals.num_direct_uses(switched) == 1 {
+                    // If there is only one direct use of `switched` we do not need to keep
+                    // it around because the only use is in the switch.
+                    Some(statement_index)
+                } else {
+                    None
+                },
+            },
+        );
+    }
+
+    if discriminants.is_empty() {
+        debug!("No SSA locals were assigned a discriminant");
+        return false;
+    }
+
+    for (switched, Discriminant { discr, block, exclude }) in discriminants {
+        let data = &body.basic_blocks[block];
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        if predecessors.iter().any(|&pred| {
+            // If we find a backedge from: `pred -> block`, this indicates that
+            // `block` is a loop header. To avoid creating irreducible CFGs we do
+            // not thread through loop headers.
+            dominators.dominates(block, pred)
+        }) {
+            debug!("Unable to thread through loop header: {block:?}");
+            continue;
+        }
+
+        let terminator = data.terminator();
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Copy(place) | Operand::Move(place),
+            targets
+        } = &terminator.kind else {
+            continue
+        };
+
+        if place.local != switched {
+            continue;
+        }
+
+        let mut finder = MutatedLocalFinder { local: discr, mutated: false };
```

Comment:
This should be created inside the loop, to ensure we don't forget to reset `mutated`.